src/charm.py

```diff
@@ -0,0 +1,347 @@
+            servers=servers,
+            server_certificate=str(SERVER_CRT_PATH),
+            server_key=str(SERVER_KEY_PATH),
+            proxy_read_timeout=self.config.get("proxy_read_timeout"),
+        )
+        self.nginx.remove_default_site()
+        self._write_nginx_logrotate_config()
+
+    def _create_server_dict(self) -> Dict[int, Set]:
+        """Create a dictionary of servers and their backends.
+
+        Returns:
+            Dict[int, Set]: A dictionary where the keys are server ports (int) and the values
+            are sets containing tuples of backends and their corresponding backend ports.
+        """
+        servers = {}
+        for request in self.load_balancer.all_requests:
+            for server_port in request.port_mapping.keys():
+                service: set = servers.setdefault(server_port, set())
+                service.update(
+                    (backend, backend_port)
+                    for backend, backend_port in itertools.product(
+                        request.backends, request.port_mapping.values()
+                    )
+                )
+        return servers
+
+    def _get_bind_addresses(self, ipv4=True, ipv6=True):
+        """Retrieve a list of bind addresses for the current unit.
+
+        Args:
+            ipv4 (bool): Whether to include IPv4 addresses (default is True).
+            ipv6 (bool): Whether to include IPv6 addresses (default is True).
+
+        Returns:
+            List[str]: A list of bind addresses available on the unit.
+        """
+        process = subprocess.run(
+            ["ip", "-j", "-br", "addr", "show", "scope", "global"],
+            capture_output=True,
+            text=True,
+            timeout=25,
+            check=True,
+        )
+        ignored_ifaces = ("lxdbr", "flannel", "cni", "virbr", "docker")
+        accept_versions = {4} if ipv4 else set()
+        accept_versions.add(6) if ipv6 else None
+
+        addrs = []
+        for addr in json.loads(process.stdout.strip()):
+            if addr["operstate"].upper() != "UP" or any(
+                addr["ifname"].startswith(prefix) for prefix in ignored_ifaces
+            ):
+                continue
+            for ifc in addr["addr_info"]:
+                local_addr = self._as_address(ifc.get("local"))
+                if local_addr and local_addr.version in accept_versions:
+                    addrs.append(str(local_addr))
+        return addrs
+
+    def _get_lb_addresses(self) -> List[str]:
+        """Return a list of load balancer addresses."""
+        forced_lb_ips = self.config.get("loadbalancer-ips").split()
+        if forced_lb_ips:
+            return forced_lb_ips
+        if self.hacluster.is_ready:
+            vips = self.config.get("ha-cluster-vip").split()
+            if vips:
+                return vips
+            dns_records = self.config.get("ha-cluster-dns").split()
+            if dns_records:
+                return dns_records
+        return []
```

Comment:
Hooray walrus (`:=`)

```suggestion
        if forced_lb_ips := self.config.get("loadbalancer-ips").split():
            return forced_lb_ips
        if self.hacluster.is_ready:
            if vips := self.config.get("ha-cluster-vip").split():
                return vips
            if dns_records := self.config.get("ha-cluster-dns").split():
                return dns_records
        return []
```